### PR TITLE
Rework Reconcile loop

### PR DESCRIPTION
**Pull request by clobrano**
Currently, the reconcile loop has two very different branches to handle
- the first reconcile loop, which ends with deletion requests
- the other reconcile loops, which assumes that the annotation existence
  means we already requested deletion.

The above assumption is not ideal, moreover it leads to a more complex
flow.

This commit reworks the loop to
- avoid assumptions on the machine deletion annotation
- only one branch flow

Signed-off-by: Carlo Lobrano <c.lobrano@gmail.com>

### CODE REVIEW

File: controllers/machinedeletionremediation_controller.go [hunk 1]
```diff
@@ -90,57 +90,68 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
-	var node *v1.Node
-	if node, err = r.getNodeFromMdr(remediation); err != nil {
+	var machine *unstructured.Unstructured
+	if machine, err = r.getMachine(machineName, machineNamespace); err != nil {
+		if apiErrors.IsNotFound(err) && !gotMachineFromNode {
+			log.Info(successfulMachineDeletionInfo, "node", remediation.Name, "machine", machineName)
+			return ctrl.Result{}, nil
+		}
+		// If Machine's name and namespace come from a Node, the Machine should still exists
+		log.Error(err, noMachineFoundError, "node", remediation.Name, "machine", machineName)
```
Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:128]: IIUC this log correspond to the use-case where apiErrors.IsNotFound(err) == true && gotMachineFromNode == true.
Does it make sence for apiErrors.IsNotFound(err) == false ?

Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:128]: IIUC this log correspond to the use-case where apiErrors.IsNotFound(err) == true && gotMachineFromNode == true. Does it make sence for apiErrors.IsNotFound(err) == false ?

File: controllers/machinedeletionremediation_controller.go [hunk 2]
```diff
@@ -90,57 +90,68 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
 
-			ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
-			if err != nil {
+	// Get Machine's Name and Namespace to retrive it from the cluster
+	var (
+		machineName, machineNamespace string
+		gotMachineFromNode            bool
+	)
+
+	if remediation.GetAnnotations() != nil {
```
[the remark this answers is not in the document]
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:101]: absolutely, I'll check for the specific annotation

File: controllers/machinedeletionremediation_controller.go [hunk 3]
```diff
@@ -90,57 +90,68 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
 				return ctrl.Result{}, err
 			}
-			if !ok {
-				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
-			}
-			return ctrl.Result{}, nil
 		}
+	} else {
+		if node, err := r.getNodeFromMdr(remediation); err != nil {
+			return ctrl.Result{}, err
```
[the remark this answers is not in the document]
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:110]: good point, I'll fix this and the other simlar occurrences

File: controllers/machinedeletionremediation_controller.go [hunk 4]
```diff
@@ -221,77 +210,100 @@ func (r *MachineDeletionRemediationReconciler) buildMachineFromNode(node *v1.Nod
...
+	phase, exists, err := unstructured.NestedString(status, "phase")
 	if err != nil {
-		return "", fmt.Errorf("could not get Machine's status.phase: error %w", err)
+		ctrl.Log.WithName("getMachineStatusPhase").Error(err, "could not get machine's status", "machine", machine.GetName())
+		return "unknown"
 	}
-	if !ok {
-		return "", fmt.Errorf("Machine object does not have a status.phase field")
+	if !exists {
+		ctrl.Log.WithName("getMachineStatusPhase").Info("machine does not have phase", "machine", machine.GetName())
```
Comment (razo7) [inline controllers/machinedeletionremediation_controller.go:301]: How about setting `ctrl.Log.WithName("getMachineStatusPhase")` once as a global var in the beginning of the file, instead of doing that multiple times in *getMachineStatusPhase* function.

machineStatusLogger := ctrl.Log.WithName("getMachineStatusPhase")

And here just use 
```suggestion
		machineStatusLogger.Info("machine does not have phase", "machine", machine.GetName())
```

**Review by razo7 — COMMENTED** [review]
Overall looks good, and I left small NITs

File: controllers/machinedeletionremediation_controller.go (hunk 4 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:301]: Do you think it would be the same to declare it locally within the function? Generally, I prefer to keep things within a smaller context whenever possible.

Comment (razo7) [reply controllers/machinedeletionremediation_controller.go:301]: > Do you think it would be the same to declare it locally within the function?

Yes, give it a try. I didn't notice that you don't use this log in other functions.. Declaring locally makes perfect sense

File: controllers/machinedeletionremediation_controller.go [hunk 5]
```diff
@@ -90,62 +90,46 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
-		return ctrl.Result{}, err
-	}
-
-	if !hasControllerOwner(machine) {
-		log.Info("ignoring remediation of node-associated machine: the machine has no controller owner", "machine", machine.GetName(), "node name", remediation.Name)
-		return ctrl.Result{}, nil
-	}
-
-	log.Info("node-associated machine found", "node", node.Name, "machine", machine.GetName())
+	if !isMachineMarkedForDeletion(machine) {
```
Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:106]: Nit: I think readability is improved by using non negative condition first.
i.e
```
if isMachineMarkedForDeletion(machine) {
...
} else {
}
```

File: controllers/machinedeletionremediation_controller.go [hunk 6]
```diff
@@ -90,62 +90,46 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
+		// save Machine's name and namespace to follow its deletion phase when its related node is deleted
+		if err = r.saveMachineNameNs(ctx, remediation, machine.GetName(), machine.GetNamespace()); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to save Machine's name and namespace")
+		}
 
-	err = r.setDeletedMachineAnnotation(ctx, remediation, machine)
-	if err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "failed to update remediation CR with timeout annotation")
+		log.Info("request node-associated machine deletion", "node", remediation.Name, "machine", machine.GetName())
+		err = r.deleteMachineOfNode(ctx, machine, remediation.Name)
```
Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:118]: `deleteMachineOfNode` is a bit confusing to me, are we verifying anywhere in the code that the machine is associated to a Node ?
In case the answer is no (maybe because such verification isn't needed due to node - machine structure)  `deleteMachine` might be a better name

File: controllers/machinedeletionremediation_controller.go [hunk 7]
```diff
@@ -90,62 +90,46 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
+			log.Info("ignoring remediation of node-associated machine: the machine has no controller owner", "machine", machine.GetName(), "node name", remediation.Name)
+			return ctrl.Result{}, nil
+		}
 
-	err = r.deleteMachineOfNode(ctx, machine, remediation.Name)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+		// save Machine's name and namespace to follow its deletion phase when its related node is deleted
+		if err = r.saveMachineNameNs(ctx, remediation, machine.GetName(), machine.GetNamespace()); err != nil {
```
Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:113]: Nit: since we are already using the MarkedForDeletion terminology, I think that  `markForDeletion` would be more suitable than `saveMachineNameNs`

File: controllers/machinedeletionremediation_controller.go (hunk 5 above)
Comment (mshitrit) [inline controllers/machinedeletionremediation_controller.go:106]: Nit:
In a very broad overview this is how I read the code (replacing if/else order):
```
if isMachineMarkedForDeletion {
 //Just do a requeue
} else{
  //1. Mark for deletion
  //2. Try to delete the machine
}
```
I think it's a bit confusing since I'd expect we will attempt to delete a Machine that is Marked for deletion.
Maybe instead of `isMachineMarkedForDeletion` something like `isMachineMarkedAsDeleted`  ?

Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:106]: I started believing that I misunderstood your original suggestion for "markedForDeletion".
The code does exactly what you outlined. 
`isMarkedForDeletion` here is a poor wrapping of checking if `deletionTimeStamp` is set. What we do is (let's say Machine still exists)

```
if Machine has not yet deletionTimeStamp
  if Machine has owner
    save Machine's data to track its deletion phase
    delete Machine
    re-queue
else
  track machine deletion phase
  re-queue
```

the `save Machine's data` step is only for logging and provide feedback to the fact that the machine is actually deleted at some point, and also necessary, otherwise when Machine's `Node` disappear, we won't be able to directly check if the Machine's still there.

File: controllers/machinedeletionremediation_controller.go (hunk 6 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:118]: Here I fell in the same problem that I "blamed" in another PR
The _node name_ is passed only for logging purpose, and it introduces a misunderstanding, so I think I will simply remove it.

Comment (slintes) [reply controllers/machinedeletionremediation_controller.go]: or log `"deleting node for remediation CR", "CR name", remediation.Name` ? 😉

File: controllers/machinedeletionremediation_controller.go (hunk 7 above)
Comment (slintes) [reply controllers/machinedeletionremediation_controller.go:115]: disagree: the name explains what the function does correctly. What we make out of it at a later stage is another story IMHO

File: controllers/machinedeletionremediation_controller.go [hunk 8]
```diff
@@ -90,67 +90,45 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
-			ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			if !ok {
-				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
-			}
+	var machine *unstructured.Unstructured
+	if machine, err = r.getMachine(remediation); err != nil {
+		if apiErrors.IsNotFound(err) {
```
Comment (slintes) [inline controllers/machinedeletionremediation_controller.go:97]: maybe add a log about this case?

File: controllers/machinedeletionremediation_controller.go [hunk 9]
```diff
@@ -90,67 +90,45 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
+		// Log deletion progress until the Machine exists
+		log.Info(postponedMachineDeletionInfo, "node", remediation.Name, "machine", machine.GetName(), "machine status.phase", getMachineStatusPhase(machine))
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
-	log.Info("node-associated machine found", "node", node.Name, "machine", machine.GetName())
+	if hasControllerOwner(machine) {
+		// save Machine's name and namespace to follow its deletion phase
+		if err = r.saveMachineNameNs(ctx, remediation, machine.GetName(), machine.GetNamespace()); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to save Machine's name and namespace")
```
Comment (slintes) [inline controllers/machinedeletionremediation_controller.go:116]: maybe add a log?

File: controllers/machinedeletionremediation_controller.go [hunk 10]
```diff
@@ -90,67 +90,45 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
-		return ctrl.Result{}, nil
+	if !machine.GetDeletionTimestamp().IsZero() {
+		// Machine deletion already requested.
+		// Log deletion progress until the Machine exists
+		log.Info(postponedMachineDeletionInfo, "node", remediation.Name, "machine", machine.GetName(), "machine status.phase", getMachineStatusPhase(machine))
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
-	log.Info("node-associated machine found", "node", node.Name, "machine", machine.GetName())
+	if hasControllerOwner(machine) {
```
Comment (slintes) [inline controllers/machinedeletionremediation_controller.go:113]: if would prefer to return early, WDYT?

```
if !hasControllerOwner {
   //log and return
}
// go on
```

File: controllers/machinedeletionremediation_controller.go [hunk 11]
```diff
@@ -90,67 +90,45 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
...
+		// save Machine's name and namespace to follow its deletion phase
+		if err = r.saveMachineNameNs(ctx, remediation, machine.GetName(), machine.GetNamespace()); err != nil {
```
Comment (slintes) [inline controllers/machinedeletionremediation_controller.go:115]: we can save an update call when we check if the annotation already exists first, either here or in saveMachineNameNs. Can happen when this call succeeds, but machine deletion below fails

File: controllers/machinedeletionremediation_controller_test.go [hunk 12]
```diff
@@ -135,26 +135,6 @@ var _ = Describe("Machine Deletion Remediation CR", func() {
 					Expect(underTest.GetAnnotations()).ToNot(BeNil())
 				})
 			})
-
-			When("machine deletion is not completed yet", func() {
```
Comment (slintes) [inline controllers/machinedeletionremediation_controller_test.go:139]: why did you remove this? :thinking:

**Review by slintes — CHANGES_REQUESTED** [review]

File: controllers/machinedeletionremediation_controller.go (hunk 5 above)
Comment (mshitrit) [reply controllers/machinedeletionremediation_controller.go]: Got it, it was my mistake :+1:

File: controllers/machinedeletionremediation_controller.go (hunk 7 above)
Comment (mshitrit) [reply controllers/machinedeletionremediation_controller.go:115]: Yep, I agree - it was my mistake of understanding the code - initially I've wrongly assumed that we use the annotation set on the remediation as a "Marker" which will later will be used indicate whether to delete the machine or not.

File: controllers/machinedeletionremediation_controller.go (hunk 6 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go]: also :smile:, but I can have the very same log in the caller as well

File: controllers/machinedeletionremediation_controller_test.go (hunk 12 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller_test.go:139]: this test was possible before this rework, because
- Reconcile loop inferred the machine deletion was requested already from the presence of the annotation
- I could trick the Reconcile loop creating a CR with an annotation inside

With the rework, this is not the case anymore, Reconcile loop always checks for Machine's existence either from CR's annotation, or Node's annotation.

To make it work, I should set a deletiontimestamp into the machine, without actually request its deletion (otherwise the test is too fast deleting it and Reconcile loop does not have the time to see the machine "deleting" phase)

File: controllers/machinedeletionremediation_controller.go (hunk 8 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:97]: the logs for this error are inside `getMachine`. I need them there, because they differ in case the Machine's NameNs comes from the annotation or from the node

Comment (slintes) [reply controllers/machinedeletionremediation_controller.go:97]: ack 👍🏼

File: controllers/machinedeletionremediation_controller_test.go (hunk 12 above)
Comment (slintes) [reply controllers/machinedeletionremediation_controller_test.go:139]: ok 👍🏼

File: controllers/machinedeletionremediation_controller.go (hunk 5 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go]: well, if the code isn't clear, it's code's problem :wink:

File: controllers/machinedeletionremediation_controller.go (hunk 10 above)
Comment (clobrano) [reply controllers/machinedeletionremediation_controller.go:113]: the idea was to create a "consistent group" of actions, but I do prefer early returns as well :)

**Comment by clobrano** [comment]
I forgot to update this part

> we can save an update call when we check if the annotation already exists first, either here or in saveMachineNameNs. Can happen when this call succeeds, but machine deletion below fails

**Comment by clobrano** [comment]
and now it's ready for review again

**Review by slintes — COMMENTED** [review]
lgtm, leaving the final review to @mshitrit since he had some comments

**Comment by mshitrit** [comment]
/lgtm

BTW - I think it would make sense to run at least one e2e test before code reviews start, that way in case some fixes are needed due to E2E failure they will be detected early and will not require a second round of reviews.

**Comment by clobrano** [comment]
> BTW - I think it would make sense to run at least one e2e test before code reviews start, that way in case some fixes are needed due to E2E failure they will be detected early and will not require a second round of reviews.

yeah, you're right
